cis194/week2/zehua/test/ExercisesSpec.hs

```diff
@@ -1,8 +1,24 @@
 module ExercisesSpec where
 
-import Test.Hspec
-import Exercises
-import Log
+import           Data.List  (sort)
+import           Exercises
+import           Log
+import           Test.Hspec
+
+-- util to make LogMessage sortable.
+-- Using `newtype` to avoid having to touch the original def in Log.hs
+newtype SortableLogMessage = SortableLogMessage { getLogMessage :: LogMessage } deriving (Eq, Show)
+instance Ord SortableLogMessage where
+  a1 <= a2 = let b1 = getLogMessage a1
+                 b2 = getLogMessage a2
+             in b1 `lteq` b2
+               where
+                 Unknown _ `lteq` _         = True
+                 _ `lteq` Unknown _         = False
+                 LogMessage _ ts1 _ `lteq` LogMessage _ ts2 _ = ts1 <= ts2
+
+sortLogMessages :: [LogMessage] -> [LogMessage]
+sortLogMessages = fmap getLogMessage . sort . fmap SortableLogMessage
```

Comment:
Is there a way to make these only visible to the test for `ex4`?

Reply:
You can add a where clause in `ex4` test to have this function defined there.

Reply:
moved into `sortLogMessages` into the test for ex4 with `let`. could not make `where` work there.

But the `newtype` and `instance` could not be moved there.

Reply:
If you have 4 spaces indentation under do notation, we can try a where clause in then end with 2 space indentation like this:
```Haskell
... $ do
    ...
  where
    a = ...
```
Using let is fine here anyway.

Reply:
Still could not make `where` work. Tried:

```
    it "combines with build should be equivalent to sort without Unknown" $ do
        let input  =
                [ (LogMessage Warning 124 "third")
                , (LogMessage Info 122 "first")
                , (LogMessage (Error 1) 123 "second")
                , (LogMessage (Error 2) 125 "fourth") ]
        (inOrder . build) input `shouldBe` sortLogMessages input
      where
        sortLogMessages :: [LogMessage] -> [LogMessage]
        sortLogMessages = fmap getLogMessage . sort . fmap SortableLogMessage
```

ghc reports error:
```
error: parse error on input ‘describe’
    |
143 |   describe "ex5" $ do
    |   ^^^^^^^^

--  While building package Week2-0.0.0 using:
      .stack/setup-exe-cache/x86_64-osx/Cabal-simple_mPHDZzAJ_2.4.0.1_ghc-8.6.5 --builddir=.stack-work/dist/x86_64-osx/Cabal-2.4.0.1 build lib:Week2 test:spec --ghc-options " -fdiagnostics-color=always"
    Process exited with code: ExitFailure 1
```

Reply:
okay. Maybe try less indent on where to be 2 space less than `it`. If this doesn’t work, I need to run this on my machines to figure out why.

Reply:
tried various things. as long as `where` is used, it is always the same complaint. 😞